File: 2025/python/benchmark_runner.py
"""
Benchmark runner for Python AOC solutions.
Runs solutions multiple times and outputs statistics to JSON.
"""
import sys
import time
import json
import statistics
import importlib
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent))

from aoc_common import read_input


def benchmark_function(func, input_data, warmup=5, iterations=20):
    """Benchmark a single function with warmup and timed runs."""
    slow_threshold_ms = 1000.0  # If iteration takes >1s, reduce iterations
    min_iterations = 3

    # Warm-up runs - check if first warmup is too slow
    for i in range(warmup):
        start = time.perf_counter()
        func(input_data)
        elapsed_ms = (time.perf_counter() - start) * 1000

        if i == 0 and elapsed_ms > slow_threshold_ms:
            iterations = min_iterations
            print(f"  First warmup took {elapsed_ms:.2f}ms, reducing iterations to {iterations}",
                  file=sys.stderr)
            break  # Skip remaining warmup iterations

    # Timed runs
    times = []
    result = None
    for i in range(iterations):
        start = time.perf_counter()
        result = func(input_data)
        end = time.perf_counter()
        elapsed_ms = (end - start) * 1000
        times.append(elapsed_ms)

        # After first timed run, check if we should reduce iterations
        if i == 0 and elapsed_ms > slow_threshold_ms and iterations > min_iterations:
            iterations = min_iterations
            print(f"  First iteration took {elapsed_ms:.2f}ms, reducing remaining iterations to {iterations-1}",
                  file=sys.stderr)

    return times, result


def benchmark_day(day, warmup=5, iterations=20):
    """Benchmark both parts of a day for all available implementations."""
    input_data = read_input(day)
    results = []

    # Try both elisp-mirrored and idiomatic implementations
    implementations = [
        ('', 'python-elisp'),              # day09.py
        ('_idiomatic', 'python-idiomatic') # day09_idiomatic.py
    ]

    for module_suffix, impl_name in implementations:
        module_name = f'day{day:02d}{module_suffix}'
        try:
            module = importlib.import_module(module_name)
        except ImportError:
            continue  # Skip if this implementation doesn't exist

        for part in [1, 2]:
            func_name = f'day{day}_part{part}'
            if hasattr(module, func_name):
                func = getattr(module, func_name)
                times, result = benchmark_function(func, input_data, warmup, iterations)

                results.append({
                    'day': day,
                    'part': part,
                    'implementation': impl_name,  # Track implementation type
                    'mean_ms': statistics.mean(times),
                    'median_ms': statistics.median(times),
                    'std_dev_ms': statistics.stdev(times) if len(times) > 1 else 0,
                    'min_ms': min(times),
                    'max_ms': max(times),
                    'iterations': len(times),
                    'result': result
                })

    return results


def run_all_benchmarks(days, warmup=5, iterations=20):
    """Run benchmarks for specified days."""
    all_results = []

    for day in days:
        print(f"Benchmarking day {day:02d}...", file=sys.stderr)
        try:
            results = benchmark_day(day, warmup, iterations)
            all_results.extend(results)
            for r in results:
                print(f"  {r['implementation']} Part {r['part']}: {r['mean_ms']:.2f}ms "
                      f"(±{r['std_dev_ms']:.2f}ms)", file=sys.stderr)
        except Exception as e:
            print(f"  Error: {e}", file=sys.stderr)
            import traceback
            traceback.print_exc(file=sys.stderr)

    return all_results


def save_results(results, filename):
    """Save benchmark results to JSON."""
    with open(filename, 'w') as f:
        json.dump(results, f, indent=2)


if __name__ == '__main__':
    days = [9, 10, 11]
    warmup = 5
    iterations = 20

    print(f"Running Python benchmarks ({warmup} warmup, {iterations} timed iterations)...",
          file=sys.stderr)
    results = run_all_benchmarks(days, warmup=warmup, iterations=iterations)

    output_file = 'python_results.json'
    save_results(results, output_file)
    print(f"\nResults saved to {output_file}", file=sys.stderr)
    print(f"Benchmarked {len(results)} solutions", file=sys.stderr)

File: 2025/lua/benchmark_runner.lua
-- Benchmark runner for LuaJIT AOC solutions
-- Runs solutions multiple times and outputs statistics to JSON

local socket = require("socket")
local json = require("cjson")
local aoc = require("aoc_common")

-- Benchmark configuration
local WARMUP = 5
local ITERATIONS = 20
local SLOW_THRESHOLD_MS = 1000.0
local MIN_ITERATIONS = 3

-- High-resolution timer
local function time_ms()
    return socket.gettime() * 1000
end

-- Statistics functions
local function mean(times)
    local sum = 0
    for _, t in ipairs(times) do
        sum = sum + t
    end
    return sum / #times
end

local function median(times)
    local sorted = {}
    for _, t in ipairs(times) do
        table.insert(sorted, t)
    end
    table.sort(sorted)
    return sorted[math.ceil(#sorted / 2)]
end

local function std_dev(times)
    if #times <= 1 then return 0 end

    local m = mean(times)
    local sum_sq = 0
    for _, t in ipairs(times) do
        sum_sq = sum_sq + (t - m) ^ 2
    end
    return math.sqrt(sum_sq / #times)
end

local function min_val(times)
    local m = times[1]
    for _, t in ipairs(times) do
        if t < m then m = t end
    end
    return m
end

local function max_val(times)
    local m = times[1]
    for _, t in ipairs(times) do
        if t > m then m = t end
    end
    return m
end

-- Benchmark a single function with warmup and timed runs
local function benchmark_function(func, input_data, warmup, iterations)
    local times = {}

    -- Warmup with adaptive iteration reduction
    for i = 1, warmup do
        local start = time_ms()
        func(input_data)
        local elapsed = time_ms() - start

        if i == 1 and elapsed > SLOW_THRESHOLD_MS then
            iterations = MIN_ITERATIONS
            io.stderr:write(string.format(
                "  First warmup took %.2fms, reducing iterations to %d\n",
                elapsed, iterations))
            break
        end
    end

    -- Timed runs
    local result
    for i = 1, iterations do
        local start = time_ms()
        result = func(input_data)
        local elapsed = time_ms() - start
        table.insert(times, elapsed)

        if i == 1 and elapsed > SLOW_THRESHOLD_MS and iterations > MIN_ITERATIONS then
            iterations = MIN_ITERATIONS
            io.stderr:write(string.format(
                "  First iteration took %.2fms, reducing remaining to %d\n",
                elapsed, iterations - 1))
        end
    end

    return times, result
end

-- Benchmark both parts of a day
local function benchmark_day(day, warmup, iterations)
    local module = require(string.format("day%02d", day))
    local input = aoc.read_input(day)
    local results = {}

    for _, part in ipairs({1, 2}) do
        local func_name = string.format("day%d_part%d", day, part)
        local func = module[func_name]

        if func then
            local times, result = benchmark_function(func, input, warmup, iterations)

            table.insert(results, {
                day = day,
                part = part,
                mean_ms = mean(times),
                median_ms = median(times),
                std_dev_ms = std_dev(times),
                min_ms = min_val(times),
                max_ms = max_val(times),
                iterations = #times,
                result = result
            })
        end
    end

    return results
end

-- Run benchmarks for specified days
local function run_all_benchmarks(days, warmup, iterations)
    local all_results = {}

    for _, day in ipairs(days) do
        io.stderr:write(string.format("Benchmarking day %02d...\n", day))
        local success, results = pcall(benchmark_day, day, warmup, iterations)

        if success then
            for _, r in ipairs(results) do
                table.insert(all_results, r)
                io.stderr:write(string.format(
                    "  Part %d: %.2fms (±%.2fms)\n",
                    r.part, r.mean_ms, r.std_dev_ms))
            end
        else
            io.stderr:write(string.format("  Error: %s\n", results))
        end
    end

    return all_results
end

-- Save benchmark results to JSON
local function save_results(results, filename)
    local f = assert(io.open(filename, "w"))
    f:write(json.encode(results))
    f:close()
end

-- Main execution
local function main()
    local days = {9, 10, 11}
    local warmup = WARMUP
    local iterations = ITERATIONS

    io.stderr:write(string.format(
        "Running LuaJIT benchmarks (%d warmup, %d timed iterations)...\n",
        warmup, iterations))

    local results = run_all_benchmarks(days, warmup, iterations)

    local output_file = "lua_results.json"
    save_results(results, output_file)

    io.stderr:write(string.format("\nResults saved to %s\n", output_file))
    io.stderr:write(string.format("Benchmarked %d solutions\n", #results))
end

-- Run if executed as script
if not pcall(debug.getlocal, 4, 1) then
    main()
end

return {
    benchmark_day = benchmark_day,
    run_all_benchmarks = run_all_benchmarks
}
